internal/translate/translate_test.go

```diff
@@ -0,0 +1,126 @@
+
+			// hp being nil means there was a fatal error in translation.
+			if hp == nil {
+				if tc.translationErrors != nil {
+					errorsDiff := cmp.Diff(translationErrors, *tc.translationErrors)
+					if errorsDiff != "" {
+						// The errors weren't what we expected.
+						t.Fatalf("Translation Errors Mismatch:\n%v", errorsDiff)
+					}
+					// httpproxy is nil, but the errors were expected, pass.
+					return
+				}
+				// Show the errors in this case, so you can set up your test case correctly.
+				t.Fatalf("HTTProxy is nil, Translation Errors:\n%v", translationErrors)
+			}
+			if tc.translated != nil {
+				translateDiff := cmp.Diff(bytes.TrimSpace(outputYAML), bytes.TrimSpace(*tc.translated))
+				if translateDiff != "" {
+					// We got a translation error
+					if tc.translationErrors != nil {
+						// We're supposed to get some errors with this translation error
+						errorsDiff := cmp.Diff(translationErrors, *tc.translationErrors)
+						if errorsDiff != "" {
+							// Translation failed, and the error wasn't what we expected.
+							t.Fatalf("\nTranslation failure:\n%v\nTranslation Errors Mismatch:\n%v", translateDiff, errorsDiff)
+						}
+					}
+					// Translation failed, we're not supposed to get any errors, log any errors we got in case.
+					t.Fatalf("\nUnexpected translation failure:\n%v, Errors:\n%v", translateDiff, translationErrors)
+
+				}
+			}
+			// Last case, if there is supposed to be no errors, and we got some, fail.
+			if tc.translationErrors == nil && len(translationErrors) > 0 {
+				t.Fatalf("Translation Errors:\n%v", translationErrors)
+			}
+		})
+	}
+}
+
+// testFixture holds test fixtures
+// The expected contract is that if there is no errors.txt, then no errors are
+// expected, and the translation should succeed.
+type testFixture struct {
+	input             []byte
+	translated        *[]byte
+	translationErrors *[]string
+}
+
+func buildFixtureSet() map[string]testFixture {
+	testdataFiles, err := ioutil.ReadDir("testdata")
+	if err != nil {
+		panic(err)
+	}
+	newFixtures := make(map[string]testFixture)
+	for _, fileinfo := range testdataFiles {
+		if !fileinfo.IsDir() {
+			continue
+		}
+
+		inputdata, err := ioutil.ReadFile(fmt.Sprintf("testdata/%s/input.yaml", fileinfo.Name()))
+		if err != nil {
+			panic(err)
+		}
+
+		newFixture := testFixture{
+			input: inputdata,
+		}
+
+		translateddata, err := ioutil.ReadFile(fmt.Sprintf("testdata/%s/output.yaml", fileinfo.Name()))
+		if err != nil {
+			// If we can't open output.yaml, then we need to signal that with a nil pointer.
+			newFixture.translated = nil
+		} else {
+			newFixture.translated = &translateddata
+		}
+
+		errordata, err := ioutil.ReadFile(fmt.Sprintf("testdata/%s/errors.txt", fileinfo.Name()))
+		if err != nil {
+			// If we can't open errors.txt, we don't have any errors defined.
+			// Signaled by nil for translationErrors.
+			newFixture.translationErrors = nil
+			newFixtures[fileinfo.Name()] = newFixture
+			continue
```

Comment:
can this condition be removed by stipulating that there should always be an errors.txt, even if it is blank, then if its missing that's a programming error on the part of the fixture writer